### PR TITLE
replace process exit with more detailed exit in src/bootstrap/*.rs

### LINKED ISSUE #98830: Don't use `process::exit` in bootstrap tests
Bootstrap has a lot of places where it calls process::exit. This makes it hard to figure out why tests fail. We should use unique exit codes so it's easier to tell where something wrong (and maybe consider using resume_unwind instead? not sure).

<details>

```
$ rg exit'\(' src/bootstrap/*.rs
src/bootstrap/util.rs
338:        std::process::exit(1);
377:        std::process::exit(1);
467:    std::process::exit(1);

src/bootstrap/toolstate.rs
96:    std::process::exit(3);
111:            std::process::exit(1);
182:            std::process::exit(1);
228:            std::process::exit(1);

src/bootstrap/tool.rs
207:                exit(1);

src/bootstrap/setup.rs
97:        std::process::exit(1);
290:            std::process::exit(1);

src/bootstrap/sanity.rs
107:            std::process::exit(1);

src/bootstrap/test.rs
676:            std::process::exit(1);
1024:                std::process::exit(1);
1254:            std::process::exit(1);

src/bootstrap/lib.rs
719:            process::exit(1);
1617:                std::process::exit(1);

src/bootstrap/format.rs
35:            std::process::exit(1);
117:        std::process::exit(1);

src/bootstrap/flags.rs
257:                process::exit(exit_code);
343:            process::exit(exit_code);
375:            process::exit(1);
596:                        std::process::exit(1);
610:                process::exit(1);
800:            process::exit(1);

src/bootstrap/compile.rs
1380:        exit(1);

src/bootstrap/config.rs
820:                    process::exit(2);
1490:        exit(1);

src/bootstrap/builder.rs
350:            std::process::exit(1);
1010:            std::process::exit(1);
1439:                    std::process::exit(1);
```

</details>

### PR DESCRIPTION
**Pull request by kons-9**
Fixes [#98830](https://github.com/rust-lang/rust/issues/98830)

I implemeted "detail_exit.rs" in lib.rs, and replace all of std::process::exit.
So, error code should panic in test code.
```
// lib.rs
pub fn detail_exit(code: i32) -> ! {
    // Successful exit
    if code == 0 {
        std::process::exit(0);
    }
    if cfg!(test) {
        panic!("status code: {}", code);
    } else {
        std::panic::resume_unwind(Box::new(code));
    }
}
```

<details>
<summary>% rg "exit\(" src/bootstrap/*.rs</summary>

```
builder.rs
351:            crate::detail_exit(1);
1000:            crate::detail_exit(1);
1429:                    crate::detail_exit(1);

compile.rs
1331:        crate::detail_exit(1);

config.rs
818:                    crate::detail_exit(2);
1488:        crate::detail_exit(1);

flags.rs
263:                crate::detail_exit(exit_code);
349:            crate::detail_exit(exit_code);
381:            crate::detail_exit(1);
602:                        crate::detail_exit(1);
616:                crate::detail_exit(1);
807:            crate::detail_exit(1);

format.rs
35:            crate::detail_exit(1);
117:        crate::detail_exit(1);

lib.rs
714:            detail_exit(1);
1620:                detail_exit(1);
1651:pub fn detail_exit(code: i32) -> ! {
1654:        std::process::exit(0);

sanity.rs
107:            crate::detail_exit(1);

setup.rs
97:        crate::detail_exit(1);
290:            crate::detail_exit(1);

test.rs
676:            crate::detail_exit(1);
1024:                crate::detail_exit(1);
1254:            crate::detail_exit(1);

tool.rs
207:                crate::detail_exit(1);

toolstate.rs
96:    crate::detail_exit(3);
111:            crate::detail_exit(1);
182:            crate::detail_exit(1);
228:            crate::detail_exit(1);

util.rs
339:        crate::detail_exit(1);
378:        crate::detail_exit(1);
468:    crate::detail_exit(1);
```
</details>

### CODE REVIEW

File: src/bootstrap/lib.rs [hunk 1]
```diff
@@ -1646,6 +1646,20 @@ fn chmod(path: &Path, perms: u32) {
 #[cfg(windows)]
 fn chmod(_path: &Path, _perms: u32) {}
 
+/// If code is not 0 (successful exit status), exit status is 101 (rust's default error code.)
+/// If the test is running and code is an error code, it will cause a panic.
+pub fn detail_exit(code: i32) -> ! {
```
Comment (jyn514) [inline src/bootstrap/lib.rs:1651]:
```suggestion
fn detail_exit(code: i32) -> ! {
```

File: src/bootstrap/lib.rs [hunk 2]
```diff
@@ -1646,6 +1646,20 @@ fn chmod(path: &Path, perms: u32) {
...
+    // Successful exit
+    if code == 0 {
```
Comment (jyn514) [inline src/bootstrap/lib.rs:1653]: Is this case ever hit? I don't think we should be using this function for successful exit, there's some cleanup in main.rs that should run. Can you instead assert the code isn't 0?

**Review by jyn514 — COMMENTED** [review]
Thanks, this is great! After you address the comments, can you squash the commits? There are instructions at https://rustc-dev-guide.rust-lang.org/git.html#advanced-rebasing.

File: src/bootstrap/lib.rs (hunk 2 above)
Comment (kons-9) [reply src/bootstrap/lib.rs:1653]: No, for example, in src/bootstrap/flags.rs , exist code.
```zsh
% rg "exit\([^1-9]" *.rs                                                                  12:34:33
flags.rs
263:                crate::detail_exit(exit_code);
349:            crate::detail_exit(exit_code);

lib.rs
1651:fn detail_exit(code: i32) -> ! {
1654:        std::process::exit(0);

% rg "exit_code" *.rs      
flags.rs
262:                let exit_code = if args.is_empty() { 0 } else { 1 };
263:                crate::detail_exit(exit_code);
331:        let usage = |exit_code: i32, opts: &Options, verbose: bool, subcommand_help: &str| -> ! {
349:            crate::detail_exit(exit_code);
```
However, such code is not much(There are 2).
Would it be better to just change the flag.rs and insert assert in the exit function?

Comment (jyn514) [reply src/bootstrap/lib.rs:1653]: Hmm, ok - seems fine to just leave it as you currently have it then.

**Comment by jyn514** [comment]
@kons-9 I see you added a new "squash the commits" commit but that didn't actually squash them. After squashing you should have exactly one commit in the PR. Make sure you use git push --force instead of merging.

**Comment by kons-9** [comment]
Sorry for the mishandling of the operation. I think this is probably done.

**Comment by jyn514** [comment]
@bors r+

Thanks for the PR!

**Pull request merged**